generator/main.cpp

```diff
@@ -53,6 +55,23 @@ void displayHelp(GeneratorSet *generatorSet);
 #include <QDebug>
 int main(int argc, char *argv[])
 {
+    /* Ensure printf function calls are always line buffered to avoid curious
+     * interleaving of the stdout and stderr streams when the output is piped
+     * to a file.  Ignore errors here (i.e. ignore closed stdout).
+     *
+     * This causes the check in test with win64_msvc2017_64 to fail.  The
+     * generator crashes here.  The cause is not known, it may be that the
+     * MSVC requires size != 0 even with a NULL buffer (ANSI-C is clear that
+     * the size is the size of the array).
+     */
+#   ifndef Q_CC_MSVC
+#       define _bufsize 0
+#   else
+#       define _bufsize 128
+#   endif
+    (void)setvbuf(stdout, NULL, _IOLBF, _bufsize);
```

Comment:
From the MS documentation for `setvbuf`:
> size
Buffer size in bytes. Allowable range: 2 <= size <= INT_MAX (2147483647). Internally, the value supplied for size is rounded down to the nearest multiple of 2.

So maybe just use a small value (like 2 or 8) for all systems.

Reply:
> This is not consistent with the ANSI-C documentation (ANSI X3.159-1989, identical to ISO C-90) wihch states in 4.9.5.6 that, "[t]he argument size specifies the size of the array." So in C-90 size should logically be 0 if the buffer pointer is NULL.

From cppstandard:
> If buffer is a null pointer, resizes the internal buffer to size. 

So I would say passing a size is fine with the buffer being NULL. Also, as we already use line buffering instead of full buffering, we could always use a sensible buffer size like 128 instead of basically using no buffering at all.

Reply:
CPP and C standards have diverged, so it's not a safe API to use.

Reply:
> CPP and C standards have diverged, so it's not a safe API to use.

The quote from cppstandard explicitely references C99, which I think we can expect to be implemented nowadays (as opposed to C90). I didn't check the standard itself, but I would expect that it is harmonized with the C++ standard as usual.

Reply:
I only have ANSI-C89 (the one I referenced) and POSIX 1003.1 which seems to defer to C90.  I found this for C99:

**Description**

[2](https://port70.net/~nsz/c/c11/n1570.html#7.21.5.6p2) The setvbuf function may be used only after the stream pointed to by stream has been associated with an open file and before any other operation (other than an unsuccessful call to setvbuf) is performed on the stream. The argument mode determines how stream will be buffered, as follows: _IOFBF causes input/output to be fully buffered; _IOLBF causes input/output to be line buffered; _IONBF causes input/output to be unbuffered. If buf is not a null pointer, the array it points to may be used instead of a buffer allocated by the setvbuf function[273)](https://port70.net/~nsz/c/c11/n1570.html#note273) and the argument size specifies the size of the array; _otherwise, size **may** determine the size of a buffer allocated by the setvbuf function_. The contents of the array at any time are indeterminate.

I suspect the ISO C++ standard will say the same thing and cppreference (which can't quote the standard verbatim) is descriptive, not authoritative.  Given that MSVC crashes, and crashes in a CVEable fashion, when it should just accept '0' I don't think setvbuf is safe for final release whereas the other changes are.

Reply:
You are right, that ("may determine the size") is indeed not as clear as I would have liked. Thanks for digging this up! I will merge it as is now.